src/vibrato.rs

```diff
@@ -0,0 +1,175 @@
+use std::f32::consts::PI;
+
+use crate::ring_buffer::RingBuffer;
+use crate::lfo::WavetableLFO;
+
+pub struct Vibrato {
+    sample_rate_hz: u32,
+    num_channels: usize,
+
+    delay_in_secs: f32,
+    oscillator_f0: f32,
+
+    delay_in_samples: usize,
+}
+
+#[derive(Debug, Clone, Copy)]
+pub enum VibratoParam {
+    OscillatorF0,
+    DelayInSecs,
+}
+
+impl Vibrato {
+    pub fn new(sample_rate_hz: u32, num_channels: usize) -> Self {
+        Vibrato {
+            sample_rate_hz: sample_rate_hz,
+            num_channels: num_channels,
+
+            delay_in_secs: f32::default(),
+            oscillator_f0: f32::default(),
+
+            delay_in_samples: usize::default(),
+        }
+    }
+
+    pub fn process(&self, input: &[&[f32]], output: &mut [&mut [f32]]) {
```

Comment:
Although it might be very hard, maybe you can try to add the block processing here.